spec/services/products_renderer_spec.rb

```diff
@@ -26,23 +26,62 @@
-      allow(distributor).to receive(:preferred_shopfront_product_sorting_method) { "by_producer" }
-      allow(distributor).to receive(:preferred_shopfront_producer_order) { "#{s2.id},#{s1.id}" }
-      products = products_renderer.send(:products)
-      expect(products).to eq([p2, p4, p1, p3])
+      it "alphabetizes products by name when taxon list is not set" do
+        allow(distributor).to receive(:preferred_shopfront_taxon_order) { "" }
+        products = products_renderer.send(:products)
+        expect(products).to eq([p1, p2, p3, p4])
+      end
     end
 
-    it "alphabetizes products by name when taxon list is not set" do
-      allow(distributor).to receive(:preferred_shopfront_taxon_order) { "" }
-      products = products_renderer.send(:products)
-      expect(products).to eq([p1, p2, p3, p4])
+    context "filtering" do
+      it "filters products by name_or_meta_keywords_or_variants_display_as_or_variants_display_name_or_supplier_name_cont" do
+        products_renderer = ProductsRenderer.new(distributor, order_cycle, customer, { q: { name_or_meta_keywords_or_variants_display_as_or_variants_display_name_or_supplier_name_cont: "abc" } })
+        products = products_renderer.send(:products)
+        expect(products).to eq([p1])
+      end
+
+      context "when property is set" do
+        let(:pa) { Spree::Property.create! name: 'A', presentation: 'A' }
+
+        it "filters products with a product property" do
+          p1.product_properties.create!({ property_id: pa.id, value: '1', position: 1 })
+          products_renderer = ProductsRenderer.new(distributor, order_cycle, customer, { q: { with_properties: [pa.id] } })
+          products = products_renderer.send(:products)
+          expect(products).to eq([p1])
+        end
+
+        it "filters products with a producer property" do
+          s1.producer_properties.create!({ property_id: pa.id, value: '1', position: 1 })
+          products_renderer = ProductsRenderer.new(distributor, order_cycle, customer, { q: { with_properties: [pa.id] } })
+          products = products_renderer.send(:products)
+          expect(products).to eq([p1, p3])
+        end
+
+        it "filters products with property when sorting is enabled" do
+          allow(distributor).to receive(:preferred_shopfront_taxon_order) { "#{t1.id},#{t2.id}" }
+          p1.product_properties.create!({ property_id: pa.id, value: '1', position: 1 })
+          p2.product_properties.create!({ property_id: pa.id, value: '1', position: 1 })
+          p3.product_properties.create!({ property_id: pa.id, value: '1', position: 1 })
+          p4.product_properties.create!({ property_id: pa.id, value: '1', position: 1 })
+          products_renderer = ProductsRenderer.new(distributor, order_cycle, customer, { q: { with_properties: [pa.id] } })
+          products = products_renderer.send(:products)
+          expect(products).to eq([p2, p4, p1, p3])
+        end
```

Comment:
This is difficult to understand. It looks like you are creating four of the same and then expect a different order. `p1` could be a "product" or a "property" and looking it up requires to go back to the top of the file.

Some ideas:

- We can give the products proper names. For example *apples*, *bananas* and so on. The alphabetical order is then clear.
- We can also name the other variables like `farmer_john` the supplier and `vegetables` the taxon and `organic` the property.
- Use a property value that gives a meaning. `A grade` would be before `organic`. Or we use `A grade` and `B grade`.

Reply:
Done !